modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/type/IgniteTypeFactory.java

```diff
@@ -255,21 +260,57 @@ else if (type instanceof IgniteCustomType)
     }
 
     /** {@inheritDoc} */
-    @Override public RelDataType leastRestrictive(List<RelDataType> types) {
+    @Override public @Nullable RelDataType leastRestrictive(List<RelDataType> types) {
         assert types != null;
         assert types.size() >= 1;
 
-        if (types.size() == 1 || allEquals(types))
+        if (types.size() == 1)
             return F.first(types);
 
+        boolean allEquals = true;
+        int uuidCnt = 0;
+        int nullCnt = 0;
+        boolean nullable = false;
+
+        for (int i = 0; i < types.size(); ++i) {
+            RelDataType type = types.get(i);
+
+            assert type != null;
+
+            if (i > 0 && allEquals && !type.equals(types.get(i - 1)))
+                allEquals = false;
+
+            if (!nullable && type.isNullable())
+                nullable = true;
+
+            if (type.getSqlTypeName() == SqlTypeName.UUID)
+                ++uuidCnt;
+            else if (type.getSqlTypeName() == SqlTypeName.NULL)
+                ++nullCnt;
+        }
+
+        if (allEquals)
+            return F.first(types);
+
+        // Least restrictive for UUID should be only NULL or UUID.
+        if (uuidCnt > 0) {
+            if (uuidCnt == types.size())
+                return createTypeWithNullability(createSqlType(SqlTypeName.NULL), nullable);
```

Comment:
Why can't we rely on Calcite implementation? Looks like this part of code never riched by tests, because SqlTypeName.NULL for UUID type is incorrect and will produce unpredictable errors. I've rolled back this changes and at least UUID tests from DataTypesTest are passed. So, I think it's better to leave default implementation.